waku-wacky-whisper-spec.md

```diff
@@ -0,0 +1,46 @@
+# Waku Wacky Whisper Specification
+
+> Version: 0.1 (Draft)
+
+> Authors: Infura "The Devil" Osaka <noreply@status.im>
+
+Basically Infura for chat. Addresses scalability issues of Whisper by sacrificing metadata protection. Specifically for resource restricted devices like mobilephones with limited data plan.
+
+インフラ (Infura) means infrastructure in Japanese, and 枠 (Waku) is a closely related word that means 'frame' or 'slide'.
+
+## Rationale
+
+Many users put a premium on bandwidth usage over metadata protection. Implementing Waku mode gives users a more performant experience, at the expense of metadata protection and requiring stronger connectivity.
+
+For more background, please see Whisper theoretical model [report](https://htmlpreview.github.io/?https://github.com/vacp2p/research/blob/master/whisper_scalability/report.html) with associated [source code](https://github.com/vacp2p/research/tree/master/whisper_scalability).
+
+## Roles and definitions
+
+1. Waku-chan
+2. Waku-san (aka Waku node)
+
+A client is said to implement waku mode if it connects to a Waku node and acts as a Waku-chan. The Waku node connects to the rest of the Whisper network.
+
+## Amendments to Whisper
+
+Waku mode is compatible with [EIP-627](https://eips.ethereum.org/EIPS/eip-627) and simply extends its capabilities. It does this through a new packet code, as well as some client specific recommendations. Two specific protocol changes are done:
+
+1. Modify EIP627 by adding a new packet code, e.g. `Topic List [101, bytes]` where bytes is a list of N topics.
```

Comment:
why not make a new protocol? what's the benefit here of trying to keep it compatible?

Reply:
To make the minimal amount of changes needed. It's only adding a capability to the current Whisper protocol, and a node can switch between the two modes. It doesn't impact the rest of the network, and they can keep using Whisper like they already are.

We can also base in on EIP-627 and call the subprotocol `whisper-waku` instead of `whisper/6`, and then simple have Waku nodes listen to whisper/6 traffic and relay it over the new subprotocol. Might be preferable?

Reply:
>  Might be preferable?

Intuitively I'd go in this direction but I haven't studied the specific details (ie what if someone else makes an "extension" to `whisper/6`, would the two be compatible? what if we want to change something about waku mode?)

Reply:
Not really clear to me what is meant here with "extension", I thought it is just extending a protocol with extra message types (new `packetId`s).

Your suggestion would mean a new subprotocol (with quite some overlap on Whisper yes, but I would not call it an extension). A Waku node would basically have to talk `whisper/6` and `whisper-waku`, and be some sort of bridge / relay between those. 

Practically, in the nim implementation a lot would be able to be reused, but there would be some extra work/complexity in sharing the message queue between protocols. Not impossible and perhaps preferable, but more work and bit more of complexity versus a better split off between protocols. Not sure if this is really that necessary though.

Reply:
We already added many packets to our Whisper fork that are not specified in the EIP-627. That path has been already taken (unfortunately) and I don't see a reason to change it now. Introducing a new packet code is waaay simpler at this stage.

Reply:
We discussed this in protocol call, and the consensus was to at least explore having a separate subprotocol. While it likely does lead to a bit of initial complexity, it's be better long term as it's clear what we are talking about when we mention a specific protocol, and it makes it easier for us to do future protocol changes, such as accounting for resources, moving to libp2p, changes to spam resistance, settlement, etc. Moving to libp2p we need a bridge between different subprotocols anyway. (If it does turn out to be impractical/take longer than expected, we can easily table it and go back to just hacking it in for the time being).

> We already added many packets to our Whisper fork that are not specified in the EIP-627. That path has been already taken (unfortunately) and I don't see a reason to change it now. Introducing a new packet code is waaay simpler at this stage.

From what I can tell that's not really documented in the current specs very clearly, e.g. ctrl-f packet https://github.com/status-im/specs/blob/master/status-whisper-usage-spec.md and https://github.com/status-im/specs/blob/master/status-client-spec.md It should be. The fact that we still seem to have undocumented behaviour that makes us not spec compliant would count as an argument in favor of more clarity on the actual protocols we support and require.

Reply:
I don't see any advantage of having a separate subprotocol that would simplify transition to libp2p honestly. If we want to have a separate subprotocol, I expect that we actually write a spec similar to EIP-627. It needs to be that comprehensive otherwise we will end up in even bigger mess than the current situation with altered Whisper handshake and additional packet codes never documented. It my opinion it is a 10x bigger task. It sounds more reasonable long term but it's just different scope.

To clarify, I guess we are not talking about a subprotocol that runs along Whisper, right? It's gonna be a completely independent self-sufficient subprotocol?

Our packet codes: https://github.com/status-im/whisper/blob/master/whisperv6/doc.go#L51-L62 vs EIP-627 packet codes: https://github.com/ethereum/go-ethereum/blob/master/whisper/whisperv6/doc.go#L48-L53.

Reply:
> spec similar to EIP-627

yep agree. A good starting point is for us to fork it and amend it.

>  It's gonna be a completely independent self-sufficient subprotocol?

yes, essentially a fork of Whisper that starts off identical. Essentially same thing we've already done except it isn't properly documented.

https://github.com/status-im/whisper/blob/master/whisperv6/doc.go#L51-L62 Why is this not documented in specs? How are we supposed to write another implementation in Nim if we just add random message types and don't document them? Not to mention other parties who are interested in building on "Whisper"/Status (e.g. team at Consensus who are evaluating it for enterprise). This is exactly what we are trying to get away from. Can we please document this in the specs repo?